### ISSUE TITLE
Since 2.3.2, we can no longer switch maintenance modes with an empty database

### ISSUE THREAD

**Issue by hostep** [magento/magento2#23577]
### Preconditions (*)
1. Magento 2.3.2
2. PHP 7.2.19

### Steps to reproduce (*)
1. Setup Magento 2.3.1 using composer
2. Drop the database and create it again so it's empty (meaning: db has zero tables)
3. Run `bin/magento maintenance:enable` or `bin/magento maintenance:disable`
4. Notice that this works => **good**
5. Setup Magento 2.3.2 using composer
6. Drop the database and create it again so it's empty (meaning: db has zero tables)
7. Run `bin/magento maintenance:enable` or `bin/magento maintenance:disable`
8. Notice that this crashes => **not good**

### Expected result (*)
1. Maintenance mode switching should work on an empty database like it did in Magento 2.3.1

### Actual result (*)
1. We get this error (including stack trace):
```
$ bin/magento maintenance:enable -vvv

In Mysql.php line 110:

  [Zend_Db_Statement_Exception (42)]
  SQLSTATE[42S02]: Base table or view not found: 1146 Table 'magento23-test.core_config_data' doesn't exist, query was: SELECT `main_table`.* FROM `core_config_data` AS `main_table`


Exception trace:
 () at vendor/magento/framework/DB/Statement/Pdo/Mysql.php:110
 Magento\Framework\DB\Statement\Pdo\Mysql->tryExecute() at vendor/magento/framework/DB/Statement/Pdo/Mysql.php:92
 Magento\Framework\DB\Statement\Pdo\Mysql->_execute() at vendor/magento/zendframework1/library/Zend/Db/Statement.php:303
 Zend_Db_Statement->execute() at vendor/magento/zendframework1/library/Zend/Db/Adapter/Abstract.php:480
 Zend_Db_Adapter_Abstract->query() at vendor/magento/zendframework1/library/Zend/Db/Adapter/Pdo/Abstract.php:238
 Zend_Db_Adapter_Pdo_Abstract->query() at vendor/magento/framework/DB/Adapter/Pdo/Mysql.php:542
 Magento\Framework\DB\Adapter\Pdo\Mysql->_query() at vendor/magento/framework/DB/Adapter/Pdo/Mysql.php:617
 Magento\Framework\DB\Adapter\Pdo\Mysql->query() at vendor/magento/zendframework1/library/Zend/Db/Adapter/Abstract.php:737
 Zend_Db_Adapter_Abstract->fetchAll() at vendor/magento/framework/Data/Collection/Db/FetchStrategy/Query.php:21
 Magento\Framework\Data\Collection\Db\FetchStrategy\Query->fetchAll() at vendor/magento/framework/Data/Collection/AbstractDb.php:778
 Magento\Framework\Data\Collection\AbstractDb->_fetchAll() at vendor/magento/framework/Data/Collection/AbstractDb.php:673
 Magento\Framework\Data\Collection\AbstractDb->getData() at vendor/magento/framework/Data/Collection/AbstractDb.php:576
 Magento\Framework\Data\Collection\AbstractDb->loadWithFilter() at vendor/magento/framework/Data/Collection/AbstractDb.php:561
 Magento\Framework\Data\Collection\AbstractDb->load() at vendor/magento/framework/Data/Collection.php:838
 Magento\Framework\Data\Collection->getIterator() at vendor/magento/module-config/App/Config/Source/RuntimeConfigSource.php:82
 Magento\Config\App\Config\Source\RuntimeConfigSource->loadConfig() at vendor/magento/module-config/App/Config/Source/RuntimeConfigSource.php:62
 Magento\Config\App\Config\Source\RuntimeConfigSource->get() at vendor/magento/framework/App/Config/ConfigSourceAggregated.php:40
 Magento\Framework\App\Config\ConfigSourceAggregated->get() at vendor/magento/module-config/App/Config/Type/System/Reader.php:60
 Magento\Config\App\Config\Type\System\Reader->read() at generated/code/Magento/Config/App/Config/Type/System/Reader/Proxy.php:95
 Magento\Config\App\Config\Type\System\Reader\Proxy->read() at vendor/magento/module-config/App/Config/Type/System.php:371
 Magento\Config\App\Config\Type\System->readData() at vendor/magento/framework/Cache/LockGuardedCacheLoader.php:79
 Magento\Framework\Cache\LockGuardedCacheLoader->lockedLoadData() at vendor/magento/module-config/App/Config/Type/System.php:261
 Magento\Config\App\Config\Type\System->loadDefaultScopeData() at vendor/magento/module-config/App/Config/Type/System.php:195
 Magento\Config\App\Config\Type\System->getWithParts() at vendor/magento/module-config/App/Config/Type/System.php:169
 Magento\Config\App\Config\Type\System->get() at vendor/magento/framework/App/Config.php:131
 Magento\Framework\App\Config->get() at vendor/magento/framework/App/Config.php:80
 Magento\Framework\App\Config->getValue() at vendor/magento/module-page-cache/Model/Config.php:119
 Magento\PageCache\Model\Config->getType() at vendor/magento/module-cache-invalidate/Observer/FlushAllCacheObserver.php:45
 Magento\CacheInvalidate\Observer\FlushAllCacheObserver->execute() at vendor/magento/framework/Event/Invoker/InvokerDefault.php:88
 Magento\Framework\Event\Invoker\InvokerDefault->_callObserverMethod() at vendor/magento/framework/Event/Invoker/InvokerDefault.php:74
 Magento\Framework\Event\Invoker\InvokerDefault->dispatch() at vendor/magento/framework/Event/Manager.php:66
 Magento\Framework\Event\Manager->dispatch() at generated/code/Magento/Framework/Event/Manager/Proxy.php:95
 Magento\Framework\Event\Manager\Proxy->dispatch() at vendor/magento/module-page-cache/Model/Cache/Type.php:51
 Magento\PageCache\Model\Cache\Type->clean() at vendor/magento/framework/App/Cache/TypeList.php:194
 Magento\Framework\App\Cache\TypeList->cleanType() at vendor/magento/framework/App/Cache/Manager.php:91
 Magento\Framework\App\Cache\Manager->clean() at vendor/magento/module-page-cache/Observer/SwitchPageCacheOnMaintenance.php:72
 Magento\PageCache\Observer\SwitchPageCacheOnMaintenance->turnOffFullPageCache() at vendor/magento/module-page-cache/Observer/SwitchPageCacheOnMaintenance.php:55
 Magento\PageCache\Observer\SwitchPageCacheOnMaintenance->execute() at vendor/magento/framework/Event/Invoker/InvokerDefault.php:88
 Magento\Framework\Event\Invoker\InvokerDefault->_callObserverMethod() at vendor/magento/framework/Event/Invoker/InvokerDefault.php:74
 Magento\Framework\Event\Invoker\InvokerDefault->dispatch() at vendor/magento/framework/Event/Manager.php:66
 Magento\Framework\Event\Manager->dispatch() at vendor/magento/framework/App/MaintenanceMode.php:82
 Magento\Framework\App\MaintenanceMode->set() at setup/src/Magento/Setup/Console/Command/AbstractMaintenanceCommand.php:90
 Magento\Setup\Console\Command\AbstractMaintenanceCommand->execute() at vendor/symfony/console/Command/Command.php:255
 Symfony\Component\Console\Command\Command->run() at vendor/symfony/console/Application.php:893
 Symfony\Component\Console\Application->doRunCommand() at vendor/symfony/console/Application.php:262
 Symfony\Component\Console\Application->doRun() at vendor/magento/framework/Console/Cli.php:105
 Magento\Framework\Console\Cli->doRun() at vendor/symfony/console/Application.php:145
 Symfony\Component\Console\Application->run() at bin/magento:23

In Mysql.php line 91:

  [PDOException (42S02)]
  SQLSTATE[42S02]: Base table or view not found: 1146 Table 'magento23-test.core_config_data' doesn't exist


Exception trace:
 () at vendor/magento/framework/DB/Statement/Pdo/Mysql.php:91
 PDOStatement->execute() at vendor/magento/framework/DB/Statement/Pdo/Mysql.php:91
 Magento\Framework\DB\Statement\Pdo\Mysql->Magento\Framework\DB\Statement\Pdo\{closure}() at vendor/magento/framework/DB/Statement/Pdo/Mysql.php:107
 Magento\Framework\DB\Statement\Pdo\Mysql->tryExecute() at vendor/magento/framework/DB/Statement/Pdo/Mysql.php:92
 Magento\Framework\DB\Statement\Pdo\Mysql->_execute() at vendor/magento/zendframework1/library/Zend/Db/Statement.php:303
 Zend_Db_Statement->execute() at vendor/magento/zendframework1/library/Zend/Db/Adapter/Abstract.php:480
 Zend_Db_Adapter_Abstract->query() at vendor/magento/zendframework1/library/Zend/Db/Adapter/Pdo/Abstract.php:238
 Zend_Db_Adapter_Pdo_Abstract->query() at vendor/magento/framework/DB/Adapter/Pdo/Mysql.php:542
 Magento\Framework\DB\Adapter\Pdo\Mysql->_query() at vendor/magento/framework/DB/Adapter/Pdo/Mysql.php:617
 Magento\Framework\DB\Adapter\Pdo\Mysql->query() at vendor/magento/zendframework1/library/Zend/Db/Adapter/Abstract.php:737
 Zend_Db_Adapter_Abstract->fetchAll() at vendor/magento/framework/Data/Collection/Db/FetchStrategy/Query.php:21
 Magento\Framework\Data\Collection\Db\FetchStrategy\Query->fetchAll() at vendor/magento/framework/Data/Collection/AbstractDb.php:778
 Magento\Framework\Data\Collection\AbstractDb->_fetchAll() at vendor/magento/framework/Data/Collection/AbstractDb.php:673
 Magento\Framework\Data\Collection\AbstractDb->getData() at vendor/magento/framework/Data/Collection/AbstractDb.php:576
 Magento\Framework\Data\Collection\AbstractDb->loadWithFilter() at vendor/magento/framework/Data/Collection/AbstractDb.php:561
 Magento\Framework\Data\Collection\AbstractDb->load() at vendor/magento/framework/Data/Collection.php:838
 Magento\Framework\Data\Collection->getIterator() at vendor/magento/module-config/App/Config/Source/RuntimeConfigSource.php:82
 Magento\Config\App\Config\Source\RuntimeConfigSource->loadConfig() at vendor/magento/module-config/App/Config/Source/RuntimeConfigSource.php:62
 Magento\Config\App\Config\Source\RuntimeConfigSource->get() at vendor/magento/framework/App/Config/ConfigSourceAggregated.php:40
 Magento\Framework\App\Config\ConfigSourceAggregated->get() at vendor/magento/module-config/App/Config/Type/System/Reader.php:60
 Magento\Config\App\Config\Type\System\Reader->read() at generated/code/Magento/Config/App/Config/Type/System/Reader/Proxy.php:95
 Magento\Config\App\Config\Type\System\Reader\Proxy->read() at vendor/magento/module-config/App/Config/Type/System.php:371
 Magento\Config\App\Config\Type\System->readData() at vendor/magento/framework/Cache/LockGuardedCacheLoader.php:79
 Magento\Framework\Cache\LockGuardedCacheLoader->lockedLoadData() at vendor/magento/module-config/App/Config/Type/System.php:261
 Magento\Config\App\Config\Type\System->loadDefaultScopeData() at vendor/magento/module-config/App/Config/Type/System.php:195
 Magento\Config\App\Config\Type\System->getWithParts() at vendor/magento/module-config/App/Config/Type/System.php:169
 Magento\Config\App\Config\Type\System->get() at vendor/magento/framework/App/Config.php:131
 Magento\Framework\App\Config->get() at vendor/magento/framework/App/Config.php:80
 Magento\Framework\App\Config->getValue() at vendor/magento/module-page-cache/Model/Config.php:119
 Magento\PageCache\Model\Config->getType() at vendor/magento/module-cache-invalidate/Observer/FlushAllCacheObserver.php:45
 Magento\CacheInvalidate\Observer\FlushAllCacheObserver->execute() at vendor/magento/framework/Event/Invoker/InvokerDefault.php:88
 Magento\Framework\Event\Invoker\InvokerDefault->_callObserverMethod() at vendor/magento/framework/Event/Invoker/InvokerDefault.php:74
 Magento\Framework\Event\Invoker\InvokerDefault->dispatch() at vendor/magento/framework/Event/Manager.php:66
 Magento\Framework\Event\Manager->dispatch() at generated/code/Magento/Framework/Event/Manager/Proxy.php:95
 Magento\Framework\Event\Manager\Proxy->dispatch() at vendor/magento/module-page-cache/Model/Cache/Type.php:51
 Magento\PageCache\Model\Cache\Type->clean() at vendor/magento/framework/App/Cache/TypeList.php:194
 Magento\Framework\App\Cache\TypeList->cleanType() at vendor/magento/framework/App/Cache/Manager.php:91
 Magento\Framework\App\Cache\Manager->clean() at vendor/magento/module-page-cache/Observer/SwitchPageCacheOnMaintenance.php:72
 Magento\PageCache\Observer\SwitchPageCacheOnMaintenance->turnOffFullPageCache() at vendor/magento/module-page-cache/Observer/SwitchPageCacheOnMaintenance.php:55
 Magento\PageCache\Observer\SwitchPageCacheOnMaintenance->execute() at vendor/magento/framework/Event/Invoker/InvokerDefault.php:88
 Magento\Framework\Event\Invoker\InvokerDefault->_callObserverMethod() at vendor/magento/framework/Event/Invoker/InvokerDefault.php:74
 Magento\Framework\Event\Invoker\InvokerDefault->dispatch() at vendor/magento/framework/Event/Manager.php:66
 Magento\Framework\Event\Manager->dispatch() at vendor/magento/framework/App/MaintenanceMode.php:82
 Magento\Framework\App\MaintenanceMode->set() at setup/src/Magento/Setup/Console/Command/AbstractMaintenanceCommand.php:90
 Magento\Setup\Console\Command\AbstractMaintenanceCommand->execute() at vendor/symfony/console/Command/Command.php:255
 Symfony\Component\Console\Command\Command->run() at vendor/symfony/console/Application.php:893
 Symfony\Component\Console\Application->doRunCommand() at vendor/symfony/console/Application.php:262
 Symfony\Component\Console\Application->doRun() at vendor/magento/framework/Console/Cli.php:105
 Magento\Framework\Console\Cli->doRun() at vendor/symfony/console/Application.php:145
 Symfony\Component\Console\Application->run() at bin/magento:23

maintenance:enable [--ip IP] [--magento-init-params MAGENTO-INIT-PARAMS]
```

### Discussion
I understand this is a weird case, but we have some script in our workflow which runs against this problem and I was a bit surprised to see this happening now in 2.3.2

**Comment by engcom-Charlie** [comment]
@magento give me 2.3-develop instance

**Comment by magento-engcom-team** [comment]
Hi @engcom-Charlie. Thank you for your request. I'm working on Magento 2.3-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @engcom-Charlie, here is your [Magento instance](https://i-23577-2-3-develop.instances.magento-community.engineering).
Admin access: [https://i-23577-2-3-develop.instances.magento-community.engineering/admin](https://i-23577-2-3-develop.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by engcom-Charlie** [comment]
Hi @hostep thank you for your report, I'm not able to reproduce issue by steps you described on clean 2.3.2-develop.
If you'd like to update issue, please reopen it.
![6](https://user-images.githubusercontent.com/51680850/60712867-9b03d400-9f20-11e9-89a2-3104b3eb3351.png)

**Comment by hostep** [comment]
@engcom-Charlie, has Magento 2.3.2 code already been merged in `2.3-develop`? Because a couple of days ago that wasn't the case yet...
This first needs to happen before you can start testing properly on the `2.3-develop` branch I think.

*Update*: confirmed, Magento 2.3.2 code hasn't been merged yet. So you should stop testing all new reported issues on `2.3-develop`, until that happens. This is slightly annoying that it takes this long for that merge to happen...

**Comment by engcom-Bravo** [comment]
Hello @hostep. Magento 2.3.2 will not merge into 2.3-develop. No plan for it.

**Comment by hostep** [comment]
Say what? Can you explain please, this makes no sense?

If you mean that the tag 2.3.2 isn't going to be merged in 2.3-develop, but all the feature branches which were merged in 2.3.2 are going to be merged in 2.3-develop, then that's fine as well, but it needs to happen pretty soon if possible!

Like how are we supposed to send in bugfixes for code which is part of 2.3.2 but not yet part of 2.3-develop? And how can you guys try to verify issue reports from Magento 2.3.2 if the code of 2.3.2 isn't merged in 2.3-develop?

**Comment by sivaschenko** [comment]
Hi @hostep can you please clarify what code from 2.3.2 is missing in 2.3-develop branch?

**Comment by hostep** [comment]
@sivaschenko: sure, I think it might be everything related to the security fixes from 2.3.2

Here are some random examples:
- [MC-5968: Fixed incorrect config override behavior](https://github.com/magento/magento2/commit/c5a083cd89f49fa5f4ec5e7f292648f6558dadb2)
- [MC-11054: Remove possibility to upload swf files through wysiwyg editor](https://github.com/magento/magento2/commit/0fdae522bcbaec4c32c653cdc2bf4a1303d2e565)
- [MC-13958: Additional Permissions for Design settings](https://github.com/magento/magento2/commit/0c8932caf5c92fa16a33531bcd063c1543cde391)
- and many many more...

**Comment by hostep** [comment]
Let's continue with the issue itself.

I just tested on 2.3-develop (using c3f29ce8efa1e59c7e1f01b6eab2f09504442924) and was able to reproduce myself, looks like @engcom-Charlie didn't test properly.

Did some git bisecting between 2.3.1 and 2.3.2, and found this commit to have caused this regression: https://github.com/magento/magento2/commit/0e373538ded9abe3051d8e9e7a04ca1b6e1d173f

Currently I have no more personal interest in seeing this fixed as we now have a workaround in our workflow scripts for this issue.

Leaving it open for now, but if Magento maintainers consider this not a regression, feel free to close.

**Comment by engcom-Charlie** [comment]
Hello @hostep ! Thank you for your contribution !
I just tested this case on 2.3-develop (using that commit) and wasn`t able to reproduce this issue . Maybe i`m missing some steps or details ? I`m closing it for now but feel free to reopen in case you can provide some extra steps/details so i will reproduce it .
Thanks in advance !

![Screenshot from 2019-07-18 10-46-18](https://user-images.githubusercontent.com/51680850/61439620-84ab3e80-a94a-11e9-9961-722c80e54940.png)

![Screenshot from 2019-07-18 10-46-53](https://user-images.githubusercontent.com/51680850/61439614-8248e480-a94a-11e9-8a02-5f937f45d1f7.png)

**Comment by hostep** [comment]
I can still reproduce on latest 2.3-develop (I've used commit 2e7f0396e37)

Steps to reproduce:
1. Clone 2.3-develop branch
2. Run `composer install`
3. Run `bin/magento setup:install ...`
4. Run following queries in mysql: `DROP DATABASE {database-name-which-was-used-in-step3}; CREATE DATABASE {database-name-which-was-used-in-step3};`
5. Run `bin/magento maintenance:enable`

Some observation I made, is that when I remove the `app/etc/env.php` file, the command works again. This now reminds me of https://github.com/magento/magento2/issues/15864, it's not exactly the same, but feels very similar.

**Comment by engcom-Charlie** [comment]
@hostep now i reproduced it . Thank you for your collaboration !

![image_2019_07_18T11_16_40_151Z](https://user-images.githubusercontent.com/51680850/61453617-496b3880-a967-11e9-907d-5920c348e31f.jpg)

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Charlie 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-18260` were created

**Issue Available**: @engcom-Charlie, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by alexadvance** [comment]
@hostep : Same issue for me whe n trying to deploy an new Magento Server with existing code.

Before 2.3.2 : env.php is present. Install fails if [install] section is present in env.php. Succeed if env.php is present but no [install] section inside.

With 2.3.2 : Install fails if env.php is present. Install succeed if env.php is absent.

**Comment by magento-engcom-team** [comment]
Hi @hostep. Thank you for your report. 
The issue has been fixed in magento/magento2#24959 by @ivan-koliadynskyy in 2.3-develop branch
Related commit(s):
 - [813f1d75628f827159f9c0496a7acff60795b53d](https://github.com/magento/magento2/commit/813f1d75628f827159f9c0496a7acff60795b53d)
 - [c252363959b2b172f48d63a5b36e7e74f02628c0](https://github.com/magento/magento2/commit/c252363959b2b172f48d63a5b36e7e74f02628c0)
 - [9e882798d0b4dca5ad9e695ddb6d7814c3e1112f](https://github.com/magento/magento2/commit/9e882798d0b4dca5ad9e695ddb6d7814c3e1112f)
 - [ee6e8a6c7e59903c49c9904b7f58797e0a7f4101](https://github.com/magento/magento2/commit/ee6e8a6c7e59903c49c9904b7f58797e0a7f4101)
 - [20e46ff09968a3a9ce1990d43c81285322eaa274](https://github.com/magento/magento2/commit/20e46ff09968a3a9ce1990d43c81285322eaa274)
 - [55a5748a744bcedddf2161b4b546c6cfd23cc2bd](https://github.com/magento/magento2/commit/55a5748a744bcedddf2161b4b546c6cfd23cc2bd)
 - [0a083be0bd2ce7b770c6da92548d3f697c7f5c01](https://github.com/magento/magento2/commit/0a083be0bd2ce7b770c6da92548d3f697c7f5c01)
 - [c579195aa7184abbe8643747736d0a2df2045d11](https://github.com/magento/magento2/commit/c579195aa7184abbe8643747736d0a2df2045d11)

The fix will be available with the upcoming 2.3.4 release.

**Comment by ilnytskyi** [comment]
@magento-engcom-team @ivan-koliadynskyy 
Can anyone check this comment ? 
https://github.com/magento/magento2/issues/26692#issuecomment-589086483

if the check result is not cached it can affect performance and increase DB load ?
